modules/imgcodecs/src/grfmt_gdal.cpp

```diff
@@ -256,35 +256,35 @@ void write_pixel( const double& pixelValue,
 
     // input: 3 channel, output: 3 channel
     else if( gdalChannels == 3 && image.channels() == 3 ){
-        if( image.depth() == CV_8U ){  image.at<Vec3b>(row,col)[channel] = newValue;  }
-        else if( image.depth() == CV_16U ){  image.ptr<Vec3s>(row,col)[channel] = newValue;  }
-        else if( image.depth() == CV_16S ){  image.ptr<Vec3s>(row,col)[channel] = newValue;  }
-        else if( image.depth() == CV_32S ){  image.ptr<Vec3i>(row,col)[channel] = newValue;  }
-        else if( image.depth() == CV_32F ){  image.ptr<Vec3f>(row,col)[channel] = newValue;  }
-        else if( image.depth() == CV_64F ){  image.ptr<Vec3d>(row,col)[channel] = newValue;  }
+        if( image.depth() == CV_8U ){  (*image.ptr<Vec3b>(row,col))[channel] = newValue;  }
+        else if( image.depth() == CV_16U ){  (*image.ptr<Vec3s>(row,col))[channel] = newValue;  }
```

Comment:
This doesn't looks as useful change. Why do you need it?

Reply:
image.ptr<Vec3b>(row,col) is not the same as image.at<Vec3b>(row,col). The image.ptr<Vec3b>() method returns a pointer to a Vec3b block, whereas image.at<Vec3b>() returns the Vec3b block itself.
In practical terms, (*image.ptr<Vec3b>(0,0))[1] will give you the 2nd byte in the data block, which is what you want, but image.ptr<Vec3b>(0,0)[1] will give you the 4th byte in the data block (the start of the 2nd Vec3b block), which is not what you want. In addition, image.ptr<Vec3b>(maxRow, maxCol)[1] will be an out-of-bounds memory access.

Reply:
The alternative fix would have been to change image.ptr<VecX>(row,col)[channel] to image.at<VecX>(row,col)[channel], but I assumed that image.ptr() was used to avoid additional bounds checks for each access that are done by image.at().

Reply:
Sorry, I noticed that the template block are missing in my comments.  Here's another try:

image.ptr&lt;Vec3b&gt;(row,col) is not the same as image.at&lt;Vec3b&gt;(row,col). The image.ptr&lt;Vec3b&gt;() method returns a pointer to a Vec3b block, whereas image.at&lt;Vec3b&gt;() returns the Vec3b block itself.
In practical terms, (*image.ptr&lt;Vec3b&gt;(0,0))[1] will give you the 2nd byte in the data block, which is what you want, but image.ptr&lt;Vec3b&gt;(0,0)[1] will give you the 4th byte in the data block (the start of the 2nd Vec3b block), which is not what you want. In addition, image.ptr&lt;Vec3b&gt;(maxRow, maxCol)[1] will be an out-of-bounds memory access.

Reply:
I see, there is corrent CV_8U case only.
BTW, `(*image.ptr<Vec3s>(row,col))` should be equal to `image.at<Vec3s>(row,col)` semantically.

Reply:
Right. The difference is that the CV_DbgAssert() calls in at() are more expensive:

```.cpp
    CV_DbgAssert((unsigned)(i1 * DataType<_Tp>::channels) < (unsigned)(size.p[1] * channels()));
    CV_DbgAssert(CV_ELEM_SIZE1(traits::Depth<_Tp>::value) == elemSize1());
```

versus this in ptr():

```.cpp
    CV_DbgAssert((unsigned)i1 < (unsigned)size.p[1]);
```

It's a performance issue rather than a semantic issue. I would note that the original code, before commit 1e9bd59f on 10/14/2015, used at() methods. This is likely where the bug was introduced in the interest of performance improvement.

Reply:
`CV_DbgAssert()` are only for debug build mode. So there is no performance question at all.

Agree #5512 introduced bug here.